### PR TITLE
feat(std): builtin bit reversing bitrev<T>(n)

### PR DESCRIPTION
**Pull request by MaxGraey**
Syntax:
```bitrev<T = u8 | u16 | u32 | u64>(n: T)```

Benchmark with different approaches: 
https://webassembly.studio/?f=xmr9vaz98sf

**DONE**
- [x] tests
- [x] definitions
- [x] portable

- [x] I've read the contributing guidelines

### CODE REVIEW

File: tests/compiler/std/polyfills.ts [hunk 1]
```diff
@@ -31,3 +31,54 @@ assert(bswap16<i16>(<i16>0xaabb) == <i16>0xbbaa);
...
+assert(bitrev<u32>(<u32>0x33333333) == <u32>0xCCCCCCCC);
+assert(bitrev<i32>(<i32>0x33333333) == <i32>0xCCCCCCCC);
+assert(bitrev<u32>(<u32>0x7FFFFFFF) == <u32>0xFFFFFFFE);
+assert(bitrev<i32>(<i32>0x7FFFFFFF) == <i32>0xFFFFFFFE);
+assert(bitrev<u32>(<u32>0x87654321) == <u32>0x84C2A6E1);
+assert(bitrev<i32>(<i32>0x87654321) == <i32>0x84C2A6E1);
+assert(bitrev<u32>(<u32>0xFFFFFFFF) == <u32>0xFFFFFFFF);
+assert(bitrev<i32>(<i32>0xFFFFFFFF) == <i32>0xFFFFFFFF);
+
+assert(false);
```
Comment (MaxGraey) [inline tests/compiler/std/polyfills.ts:84]: @dcodeIO that's pretty strange. Sometimes asserts with failures just not triggered

Comment (MaxGraey) [reply tests/compiler/std/polyfills.ts:84]: Interesting. It seems this happening only in node `v14.11.0`

Comment (MaxGraey) [reply tests/compiler/std/polyfills.ts]: Oh, that's happening earlier with SIGFAULT!

**Comment by dcodeIO** [comment]
I'm a bitrev ersed on this, in that it is neither a Wasm nor JS builtin and not ultimately required for the compiler to function, somewhat indicating that it may better be suited in user code. Do you know if there are hardware instructions that may help here, in which case perhaps proposing it as a Wasm instruction might be better?

**Comment by MaxGraey** [comment]
I think `bitreverese` is post-MVP feature similar to `bswap` and will be exposed in future (perhaps wasm 2.0?). It quite useful and clang support it btw via `__builtin_bitreverse8 / __builtin_bitreverse16 / __builtin_bitreverse32 / ...`. Also some architectures like AERM64 have single instruction for that `rbit`. `bitreverese` used in many compressed algorithms so I think will be great have something built-in especially because implement it efficiency is pretty tricky.

**Comment by lexaknyazev** [comment]
Looking at other `std/assembly/*` functions, it seems that `memory.data` is used quite extensively there. While this is perfectly fine for most users, some others (like me) would still prefer to maintain full control over their memory layout. The fact that `REV_LUT` presence depends on the compiler optimization level makes it even more complicated. At the very least, the documentation should mention the difference.

VMs, such as V8, already have internal representations for byte-swap and bit-reverse operations. Do they recognize the corresponding code patterns and replace them with native instructions like clang does sometimes? If yes, having a matching implementation in the standard library would have an additional value over keeping such functions in user's code.

> I think `bitreverse` is post-MVP feature similar to `bswap` and will be exposed in future (perhaps wasm 2.0?)

It would be great to have that proposed and added to the WebAssembly standard.

**Comment by MaxGraey** [comment]
> VMs, such as V8, already have internal representations for byte-swap and bit-reverse operations. Do they recognize the corresponding code patterns and replace them with native instructions like clang does sometimes?

No, I don't think so. Even most advanced carnelift (FF) don't do such extremely complicated pattern matching.

> Looking at other std/assembly/* functions, it seems that memory.data is used quite extensively there. While this is perfectly fine for most users, some others (like me) would still prefer to maintain full control over their memory layout

You can still use manual memory management anther manual thinks. `memory.data` just store byte sequence directly to data section and modify `__heap_base`. So if you start you manual memory from `0` better change this to `__heap_base`. That's only one minor change

**Comment by dcodeIO** [comment]
@lexaknyazev The observation is correct, yeah. We initially tried to avoid adding data segments implicitly, but this quickly fell apart when implementing lots of the math functions that are really only efficient when lookup tables are used. So at some point we decided that we cannot reasonably ensure that we don't implicit add static data segments anyway, and stopped worrying about it too much. Still, if there is a slower but more compact variant, we try to use that when compiling for size rather than for speed, though that's not always an option so basing user code upon `__heap_base` in any case is recommended.

**Pull request closed without merging**